src/dune_rules/melange_rules.ml

```diff
@@ -60,25 +66,116 @@ let build_js ~pkg_name ~module_system ~dst_dir ~cctx m =
-          [ Command.Args.S obj_dirs
-          ; melange_js_includes
-          ; As melange_package_args
-          ; A "-o"
-          ; Target output
-          ; Dep (Path.build src)
-          ]))
+  let lib_deps_js_includes = Command.Args.as_any lib_deps_js_includes in
+  Super_context.add_rule sctx ~dir:build_dir ?loc
+    (Command.run ~dir:(Path.build build_dir) (Ok compiler)
+       [ Command.Args.S obj_dirs
+       ; lib_deps_js_includes
+       ; As melange_package_args
+       ; A "-o"
+       ; Target output
+       ; Dep (Path.build src)
+       ]))
   |> Memo.Option.iter ~f:Fun.id
 
-let gen_rules ~melange_stanza_dir ~scope ~sctx ~expander
+let add_rules_for_entries ~sctx ~expander ~scope ~emit_stanza_dir ~compile_info
+    ~requires_link (mel : Melange_stanzas.Emit.t) =
+  Memo.List.iter mel.entries ~f:(function Module name ->
+      let m =
+        let basename = String.uncapitalize (Module_name.to_string name) in
+        let src_dir = Path.build emit_stanza_dir in
+        let source =
+          let impl =
+            Module.File.make Dialect.ocaml
+              (* TODO: what about .re files? *)
+              (Path.relative src_dir (basename ^ ".ml"))
+          in
+          Module.Source.make ~impl name
+        in
+        Module.of_source ~visibility:Private ~kind:Impl source
+      in
+      let obj_dir =
+        (* Maybe we should have a separate objdir for melange objs, .mobjs? *)
+        Obj_dir.make_exe ~dir:emit_stanza_dir ~name:mel.target
+      in
+      let open Memo.O in
+      let* cctx =
+        let flags = Ocaml_flags.empty in
+        let vimpl =
+          (* original impl in Lib_rules uses Virtual_rules.impl, will this break with virtual libs? *)
+          None
+        in
+        let modules = Vimpl.impl_modules vimpl (Modules.singleton m) in
+        let requires_compile = Lib.Compile.direct_requires compile_info in
+        let requires_link = Lib.Compile.requires_link compile_info in
+        let js_of_ocaml = None in
+        (* modes and pp are not passed, not sure if this will cause issues *)
+        Compilation_context.create () ~super_context:sctx ~expander ~scope
+          ~obj_dir ~modules ~flags ~requires_compile ~requires_link
+          ~opaque:Inherit_from_settings ~js_of_ocaml ~package:mel.package ?vimpl
+          ~modes:
+            { ocaml = { byte = None; native = None }
+            ; melange = Some (Requested Loc.none)
+            }
+      in
+      let pkg_name = Option.map mel.package ~f:Package.name in
+      let dst_dir = Path.Build.relative emit_stanza_dir mel.target in
+      let loc = mel.loc in
+      let lib_deps_js_includes =
+        js_includes ~emit_stanza_dir ~target:mel.target ~requires_link ~scope
+      in
+      let build_dir = (Super_context.context sctx).build_dir in
+      (* TODO: build_all should go outside Memo.List.iter, or maybe call a version of build_cm but without cctx *)
+      let* () = Module_compilation.build_all cctx in
```

Comment:
This is a big hack. I would like a way to generate cmj rules for `library` and `melange.emit` modules in a way where I can maximize the code shared, but I'm not sure how to do it yet.